instrumentation/graphql-java/graphql-java-20.0/library/src/main/java/io/opentelemetry/instrumentation/graphql/v20_0/OpenTelemetryInstrumentation.java

```diff
@@ -34,23 +77,157 @@ public InstrumentationState createState(InstrumentationCreateStateParameters par
+          for (GraphQLError error : result.getErrors()) {
+            AttributesBuilder attributes = Attributes.builder();
+            attributes.put(
+                ExceptionAttributes.EXCEPTION_TYPE, String.valueOf(error.getErrorType()));
+            attributes.put(ExceptionAttributes.EXCEPTION_MESSAGE, error.getMessage());
+
+            span.addEvent("exception", attributes.build());
+          }
+
+          executionInstrumenter.end(context, state, result, throwable);
+        });
   }
 
   @Override
   public InstrumentationContext<ExecutionResult> beginExecuteOperation(
       InstrumentationExecuteOperationParameters parameters, InstrumentationState rawState) {
+
     OpenTelemetryInstrumentationState state = ofState(rawState);
-    return helper.beginExecuteOperation(parameters, state);
+    Span span = Span.fromContext(state.getContext());
+
+    OperationDefinition operationDefinition =
+        parameters.getExecutionContext().getOperationDefinition();
+    OperationDefinition.Operation operation = operationDefinition.getOperation();
+
+    String operationType = operation.name().toLowerCase(Locale.ROOT);
+    String operationName = operationDefinition.getName();
+
+    String spanName = operationType;
+    if (operationName != null && !operationName.isEmpty()) {
+      spanName += " " + operationName;
+    }
+    span.updateName(spanName);
+
+    state.setOperation(operation);
+    state.setOperationName(operationName);
+
+    Node<?> node = operationDefinition;
+    if (sanitizeQuery) {
+      node = sanitize(node);
+    }
+    state.setQuery(AstPrinter.printAst(node));
+
+    return SimpleInstrumentationContext.noOp();
   }
 
   @Override
   public DataFetcher<?> instrumentDataFetcher(
       DataFetcher<?> dataFetcher,
       InstrumentationFieldFetchParameters parameters,
       InstrumentationState rawState) {
+
     OpenTelemetryInstrumentationState state = ofState(rawState);
-    return helper.instrumentDataFetcher(dataFetcher, state);
+
+    return environment -> {
+      ResultPath path = environment.getExecutionStepInfo().getPath();
+      Context parentContext = state.getParentContextForPath(path);
+
+      if (!dataFetcherInstrumenter.shouldStart(parentContext, environment)) {
+        // Propagate context only, do not create span
+        try (Scope ignored = parentContext.makeCurrent()) {
+          return dataFetcher.get(environment);
+        }
+      }
+
+      if (parameters.isTrivialDataFetcher() && !createSpansForTrivialDataFetcher) {
+        // Propagate context only, do not create span
+        try (Scope ignored = parentContext.makeCurrent()) {
+          return dataFetcher.get(environment);
+        }
+      }
+
+      // Start span
+      Context childContext = dataFetcherInstrumenter.start(parentContext, environment);
+      state.setContextForPath(path, childContext);
+
+      boolean isCompletionStage = false;
+
+      try {
+        Object fieldValue = dataFetcher.get(environment);
```

Comment:
> My understanding is the context was made current when I called;
> Context childContext = dataFetcherInstrumenter.start(parentContext, environment);

calling `start` does not make the returned context current you have to use `makeCurrent`

> Also the unit tests I've written seem to confirm the code is working.

your tests don't create child spans inside data fetchers, if they did, like the `AbstractGraphqlTest` does, then the test would fail

Reply:
You are correct. I think I finally understand how `dataFetcherInstrumenter.start(parentContext, environment);` works. It just starts the span but does not set the context.

I've since updated my code and unit tests to make sure this is working as expected.